exercises/edward_brown/ex01_basics/include/ForEachWord.h

```diff
@@ -0,0 +1,29 @@
+#ifndef EX1_FOR_EACH_WORD
+#define EX1_FOR_EACH_WORD
+#include "DiscardWhile.h"
+#include "TakeWhile.h"
+#include "WordCharPredicates.h"
+
+/** Calls the DoIt action on all words found in the range [Begin, End).
+ *  Here a word is a sequence of one or more whitespace characters terminated
+ *  by End or a whitespace character.
+ *
+ *  @param Begin The const_iterator marking the begining of the range in which
+ * to find words.
+ *  @param End The const_iterator marking the end of the range in which to find
+ * words.
+ *  @param DoIt A callable object to be called for each word found in the range
+ * [Begin, End).
+ */
+template <typename ConstCharInputIterator, typename Action>
+void forEachWord(ConstCharInputIterator Begin, ConstCharInputIterator End,
+                 Action DoIt) {
+  auto Current = Begin;
+  while (Current != End) {
+    auto WordPositionPair = takeWhile(Current, End, &isWordChar);
+    DoIt(std::move(WordPositionPair.first));
```

Comment:
If `DoIt` wished to modify the lifetime of the string, keep a copy or wished to modify it without affecting the caller then there may be a benefit in the form of a move instead of a copy. But I can't remember the exact details regarding the implementation of `std::unordered_map` so I changed to by reference passing instead. See 622cf98fa6936e055d3aa341a2a2522dc32aa6ff